File: python/src/ancilis/producers/auto.py
"""Auto-detection of installed LLM/framework SDKs.

Removes per-SDK boilerplate from user code: call ``auto_register(config,
engine)`` and get back a dict of producers keyed by provider, one per
upstream SDK present in the current environment.

This module never *imports* the upstream SDKs — it uses
``importlib.util.find_spec`` so detection is free of side effects.

Typical wiring::

    from ancilis import Engine, load_config
    from ancilis.producers.auto import auto_register

    config = load_config("ancilis.yaml")
    engine = Engine(config)
    producers = auto_register(config, engine)
    # producers == {"anthropic": AnthropicActionProducer(...), ...}

For UI/diagnostics use ``detect_installed_sdks()`` which returns a
plain dict ``{provider_slug: True/False}`` without touching any state.
"""

from __future__ import annotations

import importlib.util
from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

from ancilis.config import ResolvedConfig
from ancilis.engine.engine import Engine
from ancilis.engine.registry import ToolRegistry
from ancilis.evidence.store import EvidenceStore


@dataclass(frozen=True)
class _Detector:
    """Maps a producer slug to the upstream module(s) that signal availability."""

    provider: str
    producer_attr: str  # name on ancilis.producers package
    modules: tuple[str, ...]  # any-of: if any module is importable, the SDK is present


# Each row: (provider_slug, producer_class_attr, candidate_module_names...)
# A provider is considered "available" if importlib can locate any one of
# its candidate modules. Aliased modules cover renames (e.g. google.genai vs
# google.generativeai, autogen vs autogen_agentchat vs ag2).
_DETECTORS: tuple[_Detector, ...] = (
    _Detector("anthropic", "AnthropicActionProducer", ("anthropic",)),
    _Detector("openai", "OpenAIActionProducer", ("openai",)),
    _Detector("gemini", "GeminiActionProducer", ("google.genai", "google.generativeai")),
    _Detector("mistral", "MistralActionProducer", ("mistralai",)),
    _Detector("cohere", "CohereActionProducer", ("cohere",)),
    _Detector("groq", "GroqActionProducer", ("groq",)),
    _Detector("together", "TogetherActionProducer", ("together",)),
    _Detector("fireworks", "FireworksActionProducer", ("fireworks",)),
    _Detector("aws-bedrock", "BedrockActionProducer", ("boto3",)),
    _Detector(
        "langchain",
        "LangChainActionProducer",
        ("langchain", "langchain_core"),
    ),
    _Detector("crewai", "CrewAIActionProducer", ("crewai",)),
    _Detector(
        "autogen",
        "AutoGenActionProducer",
        ("autogen", "autogen_agentchat", "ag2"),
    ),
    _Detector(
        "semantic-kernel",
        "SemanticKernelActionProducer",
        ("semantic_kernel",),
    ),
)


def _module_present(module_names: Iterable[str]) -> bool:
    """True iff any candidate module is importable (no actual import)."""
    for name in module_names:
        try:
            spec = importlib.util.find_spec(name)
        except (ImportError, ValueError):
            spec = None
        if spec is not None:
            return True
    return False


def detect_installed_sdks() -> dict[str, bool]:
    """Return a flat ``{provider_slug: present}`` dict.

    Side-effect-free; safe to call before any producer is wired up. Useful
    for CLI diagnostics ("which SDKs am I going to wrap?") and for users who
    want to make scope decisions before calling ``auto_register``.
    """
    return {d.provider: _module_present(d.modules) for d in _DETECTORS}


def installed_provider_slugs() -> list[str]:
    """Return only the provider slugs whose upstream SDK is installed."""
    return [provider for provider, present in detect_installed_sdks().items() if present]


def _instantiate_producer(
    detector: _Detector,
    *,
    config: ResolvedConfig,
    engine: Engine,
    registry: ToolRegistry | None,
    evidence_store: EvidenceStore | None,
) -> Any:
    """Resolve the producer class via the lazy ``ancilis.producers`` package
    and instantiate it with the standard producer constructor signature."""
    from ancilis import producers as producers_pkg

    cls = getattr(producers_pkg, detector.producer_attr)
    return cls(
        config=config,
        engine=engine,
        registry=registry,
        evidence_store=evidence_store,
    )


def auto_register(
    config: ResolvedConfig,
    engine: Engine,
    *,
    registry: ToolRegistry | None = None,
    evidence_store: EvidenceStore | None = None,
    include: Iterable[str] | None = None,
    exclude: Iterable[str] | None = None,
) -> dict[str, Any]:
    """Instantiate one producer per detected upstream SDK.

    Returns a dict keyed by provider slug. Filters:
    - ``include``: only consider these provider slugs (still must be installed).
    - ``exclude``: skip these provider slugs even if installed.

    The producers share the engine and registry, and unless an explicit
    ``evidence_store`` is passed, each producer instantiates its own (matching
    the existing single-producer wiring patterns).
    """
    include_set = set(include) if include is not None else None
    exclude_set = set(exclude) if exclude else set()

    available = detect_installed_sdks()
    out: dict[str, Any] = {}
    for detector in _DETECTORS:
        if not available.get(detector.provider, False):
            continue
        if include_set is not None and detector.provider not in include_set:
            continue
        if detector.provider in exclude_set:
            continue
        out[detector.provider] = _instantiate_producer(
            detector,
            config=config,
            engine=engine,
            registry=registry,
            evidence_store=evidence_store,
        )
    return out

File: typescript/src/ancilis/producers/auto.ts
/**
 * Auto-detection of installed LLM/framework SDKs (TypeScript parity with
 * `ancilis.producers.auto`).
 *
 * Detects which upstream npm packages are installed in the current
 * environment and instantiates one producer per detected SDK. Uses
 * `createRequire(import.meta.url).resolve(...)` — never actually imports
 * the upstream package, so detection has no side effects.
 *
 * Typical wiring:
 *
 *   import { loadConfig } from "ancilis";
 *   import { Engine } from "ancilis";
 *   import { autoRegister } from "ancilis";
 *
 *   const config = loadConfig({ raw: { agent: { name: "my-agent" } } });
 *   const engine = new Engine(config);
 *   const producers = autoRegister(config, engine);
 *   // producers == { anthropic: AnthropicActionProducer, openai: ..., ... }
 */

import { createRequire } from "node:module";
import type { ResolvedConfig } from "../config/index.js";
import { Engine } from "../engine/engine.js";
import { ToolRegistry } from "../engine/registry.js";
import { EvidenceStore } from "../evidence/store.js";
import {
  AnthropicActionProducer,
  CohereActionProducer,
  DeepSeekActionProducer,
  FireworksActionProducer,
  GeminiActionProducer,
  GroqActionProducer,
  type LLMActionProducer,
  MistralActionProducer,
  OpenAIActionProducer,
  TogetherActionProducer,
  XAIActionProducer,
} from "./llm.js";
import { BedrockActionProducer } from "./bedrock.js";
import { LangChainActionProducer } from "./langchain.js";
import { CrewAIActionProducer } from "./crewai.js";
import { AutoGenActionProducer } from "./autogen.js";
import { SemanticKernelActionProducer } from "./semantic_kernel.js";

const require_ = createRequire(import.meta.url);

type ProducerCtor = new (
  config: ResolvedConfig,
  engine: Engine,
  registry?: ToolRegistry,
  evidenceStore?: EvidenceStore,
) =>
  | LLMActionProducer
  | BedrockActionProducer
  | LangChainActionProducer
  | CrewAIActionProducer
  | AutoGenActionProducer
  | SemanticKernelActionProducer;

interface Detector {
  /** Slug used as the dict key in the result. */
  provider: string;
  /** Producer class to instantiate when this SDK is detected. */
  cls: ProducerCtor;
  /** Any-of: SDK is "available" if any candidate npm package is resolvable. */
  modules: readonly string[];
}

// SDK → producer mapping. Aliases cover renames (e.g. @google/genai vs
// @google/generative-ai). Empty modules tuple = no dedicated SDK; user
// must wire explicitly. Bedrock keys off the AWS SDK v3 runtime client.
const DETECTORS: readonly Detector[] = [
  { provider: "anthropic", cls: AnthropicActionProducer, modules: ["@anthropic-ai/sdk"] },
  { provider: "openai", cls: OpenAIActionProducer, modules: ["openai"] },
  {
    provider: "gemini",
    cls: GeminiActionProducer,
    modules: ["@google/genai", "@google/generative-ai"],
  },
  { provider: "mistral", cls: MistralActionProducer, modules: ["@mistralai/mistralai"] },
  { provider: "cohere", cls: CohereActionProducer, modules: ["cohere-ai"] },
  { provider: "groq", cls: GroqActionProducer, modules: ["groq-sdk"] },
  { provider: "together", cls: TogetherActionProducer, modules: ["together-ai"] },
  { provider: "fireworks", cls: FireworksActionProducer, modules: ["fireworks-ai"] },
  { provider: "deepseek", cls: DeepSeekActionProducer, modules: [] },
  { provider: "xai", cls: XAIActionProducer, modules: [] },
  {
    provider: "aws-bedrock",
    cls: BedrockActionProducer,
    modules: ["@aws-sdk/client-bedrock-runtime", "aws-sdk"],
  },
  {
    provider: "langchain",
    cls: LangChainActionProducer,
    modules: ["@langchain/core", "langchain"],
  },
  { provider: "crewai", cls: CrewAIActionProducer, modules: ["crewai"] },
  { provider: "autogen", cls: AutoGenActionProducer, modules: ["autogen"] },
  {
    provider: "semantic-kernel",
    cls: SemanticKernelActionProducer,
    modules: ["@semantic-kernel/typescript", "@microsoft/semantic-kernel"],
  },
];

/** Internal: returns true iff any candidate module name resolves. */
export function _modulePresent(modules: readonly string[]): boolean {
  for (const name of modules) {
    try {
      require_.resolve(name);
      return true;
    } catch {
      // not installed; continue
    }
  }
  return false;
}

export interface AutoRegisterOptions {
  registry?: ToolRegistry;
  evidenceStore?: EvidenceStore;
  /** Only consider these provider slugs (still must be installed). */
  include?: readonly string[];
  /** Skip these provider slugs even if installed. */
  exclude?: readonly string[];
}

/** Side-effect-free detection: returns a flat `{provider: present?}` map. */
export function detectInstalledSdks(): Record<string, boolean> {
  const out: Record<string, boolean> = {};
  for (const d of DETECTORS) {
    out[d.provider] = d.modules.length > 0 ? _modulePresent(d.modules) : false;
  }
  return out;
}

/** Slugs whose upstream SDK is detected as installed. */
export function installedProviderSlugs(): string[] {
  const detected = detectInstalledSdks();
  return Object.keys(detected).filter((p) => detected[p] === true);
}

/**
 * Instantiate one producer per detected upstream SDK. Returns a record
 * keyed by provider slug.
 */
export function autoRegister(
  config: ResolvedConfig,
  engine: Engine,
  options: AutoRegisterOptions = {},
): Record<
  string,
  | LLMActionProducer
  | BedrockActionProducer
  | LangChainActionProducer
  | CrewAIActionProducer
  | AutoGenActionProducer
  | SemanticKernelActionProducer
> {
  const include = options.include ? new Set(options.include) : null;
  const exclude = new Set(options.exclude ?? []);
  const out: Record<string, LLMActionProducer | BedrockActionProducer | LangChainActionProducer | CrewAIActionProducer | AutoGenActionProducer | SemanticKernelActionProducer> = {};
  for (const d of DETECTORS) {
    if (d.modules.length === 0) continue;
    if (!_modulePresent(d.modules)) continue;
    if (include !== null && !include.has(d.provider)) continue;
    if (exclude.has(d.provider)) continue;
    out[d.provider] = new d.cls(config, engine, options.registry, options.evidenceStore);
  }
  return out;
}

// Test-only export for the detector table integrity check.
export { DETECTORS as _DETECTORS };
